CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen

veronikaslc opened the pull request.

Commit:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen

veronikaslc commented:
TODO: Fix functionality of the Form page.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/PrintButton.jsx:
```diff
@@ -59,6 +59,25 @@ function PrintButton(props) {
+    event.stopPropagation();
+    event.preventDefault ? event.preventDefault() : (event.returnValue = false);
+
+    if (event.ctrlKey && (event.key == "p" || event.keyCode == 80)) {
```
Let's also try to catch MacOS specific shortcuts.

Commit:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Added meta key option for detection Mac keyboard print ⌘ + P keydown events

sashaandjic commented:
Crtl P seems to open 2 previews

sdumitriu reviewed (dismissed):
This stops all key events when the Print button is visible, not just Ctrl+P, but also Ctrl+R, Up and Down... And it doesn't stop the native dialog when the button is not visible, for example open a form in view mode, CtrlP is not stopped while the action menu is collapsed.

veronikaslc commented:
Yup, thats why there is a TODO: Fix functionality of the Form page.

Commit:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Stop propagation only for the printing keydown events

Commit:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Made actions menu hidden instead of absent to enable PrintButton listeners attach for printing keydown

veronikaslc commented:
TODO: Crtl + P seems to open 2 previews

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/PrintButton.jsx:
```diff
@@ -59,6 +59,24 @@ function PrintButton(props) {
+      // unsubscribe event
+      document.removeEventListener("keydown", handleOnPrintKeydown);
+    };
+  });
```
I think this should solve the double dialog open:

```suggestion
  }, []);
```

veronikaslc replied:
`ResourceHeader` duplicates `PrintButton` when renders `actions` 2 times for extended and collapsed header modes for scrolling down, so there are actually 2 previews that are getting open.

marta- replied:
@sdumitriu Nope. While this change should indeed be made to avoid adding and removing the listener unnecessarily with state changes, the double print dialog is caused by the fact that the `ResourceHeader` component renders the menu twice: once next to the expanded title when the full header is displayed, and once in the collapsed version. We have to either render the menu on;y once or attach the listener only once.

veronikaslc committed:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen

Avoid unnecessary listeners addition/removal

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

marta- committed:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Fixed diplication of Print preview dialog by changing ResourceHeader to not render the menu twice

marta- commented:
> TODO:
> 
> * Crtl + P seems to open 2 previews

Fixed in 99387b7 .

Commit:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Fixed the Actions menu popoup that was breaking the scroll

veronikaslc commented:
Fixed 
- Popover hidden mode breaks scroll in https://github.com/data-team-uhn/cards/pull/900/commits/51fb22333d4a5bf2f4105140fad409dc89e2ea3d

marta- committed:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Restored the ability to close the menu by clicking outside

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (changes requested):
Multiple dialogs can still pop up:
- create a Patient with a couple of visits
- open the patient chart
- press Ctrl+P
- a dialog pops up for each Visit subject, and one for the Patient subject itself

marta- committed:
CARDS-1617: Ctrl+P bypasses react-to-print and prints what is on the screen
Bug fix: Ctrl+P no longer activates the print buttons of child subjects on a subject page

marta- commented:
> Multiple dialogs can still pop up:
> 
> * create a Patient with a couple of visits
> * open the patient chart
> * press Ctrl+P
> * a dialog pops up for each Visit subject, and one for the Patient subject itself

Should be fixed in d8e4b85 .

sashaandjic commented:
> Multiple dialogs can still pop up:
> 
> * create a Patient with a couple of visits
> * open the patient chart
> * press Ctrl+P
> * a dialog pops up for each Visit subject, and one for the Patient subject itself

Reproduced

marta- commented:
> > Multiple dialogs can still pop up:
> > 
> > * create a Patient with a couple of visits
> > * open the patient chart
> > * press Ctrl+P
> > * a dialog pops up for each Visit subject, and one for the Patient subject itself
> 
> Reproduced

@sashaandjic Does this mean it still happens after the latest commit?

sashaandjic commented:
> @sashaandjic Does this mean it still happens after the latest commit?

No. I just re-tested it the latest build. I reproduced it before the latest change. All good.
